### PR TITLE
TST: Add cygwin build to CI

### PR DESCRIPTION
**Pull request by DWesl**
Attempt to add a Cygwin build to the CI.  

The code builds just fine, but the actual "running the tests" bit is crashing for reasons I don't understand and can't reproduce on my local machine.  One way around this would be to build a wheel, install that into a virtual environment, and test using `np.test()`.  Another would be using `tox` to run the tests, basically automating the previous suggestion.

Inspired by discussion on #18308.

### CODE REVIEW

**Comment by DWesl** [comment]
I suppose I should mention: `python runtests.py` gets through the build process fine, and I'm assuming it starts the testing process because it breaks when trying to import `numpy.linalg` because it can't find `lapack_lite` or `_umath_linalg`.  I'm assuming the other CI has a way of setting things up so those modules are found, but I'm not sure what that is.

I've been using `python runtests.py` from the NumPy root directory for testing on Cygwin for a while now without issues at this point.  I'm not sure what's going on here, and have not been able to reproduce this locally.

**Comment by mattip** [comment]
I wonder what is up with the doc build?
```
rm -rf build/*
find . -name generated -type d -prune -exec rm -rf "{}" ";"
# for testing
# @echo installed numpy 4d148aa1a9 matches git version 4d148aa1a9; exit 1
mkdir -p build
touch build/generate-stamp
mkdir -p build/html build/doctrees
LANG=C sphinx-build -b html -WT --keep-going -d build/doctrees  -q source build/html 
1.21.dev0 1.21.0.dev0+677.g4d148aa1a
make: *** [Makefile:179: html-build] Terminated

Too long with no output (exceeded 10m0s): context deadline exceeded
```

**Comment by DWesl** [comment]
> I wonder what is up with the doc build?
> 
> ```bash
> LANG=C sphinx-build -b html -WT --keep-going -d build/doctrees  -q source build/html 
> 1.21.dev0 1.21.0.dev0+677.g4d148aa1a
> make: *** [Makefile:179: html-build] Terminated
> 
> Too long with no output (exceeded 10m0s): context deadline exceeded
> ```

No idea.  I'm hoping it's a transient thing that goes away with the new commit.

I tried building a NumPy wheel, installing it, and testing against that, but is still failed to import `numpy`.  I think the new version will get the run status to reflect that.  I'm still not sure why this isn't working, since those are the steps used to generate the NumPy I'm currently using on Cygwin.  I'm running rather low on ideas at this point.

**Comment by seberg** [comment]
It has been going on a lot, I have a PR open to fix it (sorry missed the discussion here).  My theory is that we are getting different hardware so that it now takes sometimes up to 30 instead of around 10 minutes, and so we suddenly get timeouts frequently.  In any case, seems that bumping the limit in gh-18349 should make it reliable again.

**Comment by DWesl** [comment]
The test is failing with "ImportError: No such file or directory" after attempting to import numpy.linalg.lapack_lite.  A previous step said the file it's trying to open (`/usr/local/lib/python3.8/site-packages/numpy/linalg/lapack_lite.cpython-38-x86_64-cygwin.dll`) exists and has the right permissions (read and execute for everyone).  

I could try a user install, a virtual environment, or tox, but I'm not sure those would make a difference.  I'm still not sure what's going on here.

**Comment by mattip** [comment]
Unfortunately, windows does not have tooling like Linux to detect which DLLs are missing when loading a DLL at runtime. Does cygwin provide an `ldd`-like command to do anything like that? It might be something around the fortran or openblas dll.

**Comment by DWesl** [comment]
> Unfortunately, windows does not have tooling like Linux to detect which 
> DLLs are missing when loading a DLL at runtime. Does cygwin provide an 
> |ldd|-like command to do anything like that? It might be something 
> around the fortran or openblas dll.
> 

It provides |ldd| and an alternate command I'm more familiar with.  I 
added both to the check on the dlls.

**Comment by DWesl** [comment]
`ldd` reports no missing dependencies for any of the NumPy-installed DLLs as reported by pip.  All of the DLLs also have execute permissions for all users, and report no attributes that would change the permissions from 0755.  

Is the failure to load something specific to Windows or Docker or the combination?

**Comment by charris** [comment]
> Is the failure to load something specific to Docker or Docker running Windows?

I don't know, but Windows has been tricky in that regard in the past. What we need here is some input from a Windows person.

**Comment by DWesl** [comment]
> > Is the failure to load something specific to Docker or Docker running Windows?
> 
> I don't know, but Windows has been tricky in that regard in the past. What we need here is some input from a Windows person.

Is there some flag we need to set to tell Docker or Windows that we want to use this file as a DLL, despite it being created in this container?

**Comment by charris** [comment]
Python 3.8 changed the way load paths were specified on Windows, so that might be related.  See ([permalink](http://numpy-discussion.10968.n7.nabble.com/Compile-NumPy-with-ifort-MSVC-and-MKL-DLL-load-failed-tp48968p48970.html)). @cgohlke Thoughts?

**Comment by DWesl** [comment]
## A bit of background on Cygwin BLAS and LAPACK
The netlib reference BLAS and LAPACK dlls are installed to `/usr/lib/lapack`.  The OpenBLAS dll is installed to `/usr/bin` under the same name as the reference BLAS dll.  The default setup puts `/usr/bin` before `/usr/lib/lapack` in the `PATH`, so Windows finds OpenBLAS and loads that instead of the reference BLAS.

## Attempts to load BLAS and LAPACK from Python3.8 on my local Cygwin
```bash
$ ls /usr/bin/cygwin1.dll
/usr/bin/cygwin1.dll

$ python3.8
Python 3.8.7 (default, Jan 31 2021, 21:50:45)
[GCC 10.2.0] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
```
```python
>>> import os
>>> os.add_dll_directory
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'os' has no attribute 'add_dll_directory'
>>> import ctypes.util
>>> ctypes.CDLL("cygwin1.dll").cygwin_conv_path
<_FuncPtr object at 0x6fffffe6eac0>
>>> ctypes.CDLL(ctypes.util.find_library("lapack")).dgesv_
<_FuncPtr object at 0x6ffff790dd00>
>>> import numpy.linalg.lapack_lite
>>>
```
The lack of `os.add_dll_directory` on  Cygwin implies the DLL load path didn't change.
If `/usr/lib/lapack` not being on the rewritten DLL load path list were an issue, one of the LAPACK loads would fail.  The load issue might be that `/usr/lib/lapack` isn't on the PATH, but that would probably show up in the `ldd` output.  

I could try switching to 3.7 if you think that would help.

**Comment by charris** [comment]
> I could try switching to 3.7 if you think that would help.

It would eliminate one possible concern. Have you tried searching stackoverflow?

**Comment by DWesl** [comment]
I tried with python 3.7:
```
ls
-rwxr-xr-x 1 runneradmin None   194903 Feb 18 18:52 /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll
file
/usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll:       PE32+ executable (DLL) (console) x86-64, for MS Windows
ldd
/usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll:
	ntdll.dll => /cygdrive/c/windows/SYSTEM32/ntdll.dll (0x7ffd59240000)
	KERNEL32.DLL => /cygdrive/c/windows/System32/KERNEL32.DLL (0x7ffd574e0000)
	KERNELBASE.dll => /cygdrive/c/windows/System32/KERNELBASE.dll (0x7ffd560b0000)
	lapack_lite.cpython-37m-x86_64-cygwin.dll => /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll (0x5cde30000)
	libpython3.7m.dll => /usr/bin/libpython3.7m.dll (0x3fc150000)
	cygwin1.dll => /usr/bin/cygwin1.dll (0x180040000)
cygcheck
C:\tools\cygwin\lib\python3.7\site-packages\numpy\linalg\lapack_lite.cpython-37m-x86_64-cygwin.dll
  C:\tools\cygwin\bin\cygwin1.dll
    C:\windows\system32\KERNEL32.dll
      C:\Program Files\PowerShell\7\api-ms-win-core-rtlsupport-l1-1-0.dll
      C:\windows\system32\ntdll.dll
      C:\windows\system32\KERNELBASE.dll
        C:\Program Files\PowerShell\7\api-ms-win-eventing-provider-l1-1-0.dll
      C:\Program Files\PowerShell\7\api-ms-win-core-processthreads-l1-1-0.dll
      ...
      C:\Program Files\PowerShell\7\api-ms-win-core-psapi-ansi-l1-1-0.dll
  C:\tools\cygwin\bin\libpython3.7m.dll
    C:\tools\cygwin\bin\cygintl-8.dll
      C:\tools\cygwin\bin\cygiconv-2.dll
    C:\tools\cygwin\bin\cyggcc_s-seh-1.dll

python -vv
# trying /usr/lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-x86_64-cygwin.dll
Traceback (most recent call last):
    ...
ImportError: No such file or directory
```
Same error.  None of my searches ("GitHub Actions on Windows", "Problems loading DLLs in GitHub Actions on Windows", or "GitHub Actions on Windows third-party DLL") returned anything relevant.

**Comment by mattip** [comment]
Could you save the wheel as a [github action artifacts](https://github.com/actions/upload-artifact)and then download that to a local machine to test it? Maybe something is missing in the packaging

**Comment by DWesl** [comment]
I installed a snapshot of the Cygwin dll a few months back to get around a segfault in `modfl`.  This means `uname -r` starts with `3.2.0` rather than the `3.1.7` of the current release.  As a consequence, pip doesn't recognize the uploaded wheel as valid for my platform (it also doesn't recognize wheels built with Cygwin 3.1.6 as valid on 3.1.7, but that's not relevant here).

Does anyone know how to get pip to ignore this and install the wheel anyway?  I got `numpy.linalg.lapack_lite` to import by unpacking with `wheel` then adding execute permissions to all installed dlls.  `np.linalg.test()` passes, so loading `lapack_lite` seems to work, narrowing the problem to some difference between my system and the GitHub runners.

**Comment by DWesl** [comment]
I seem to have a similar problem with a different project on my computer.  I still have no idea why it's not importing the file it finds with the proper permissions, but I can investigate this locally now.

**Comment by charris** [comment]
Great!

**Comment by DWesl** [comment]
Thanks @mattip for suggesting I look at `ldd` output.  I eventually saw problems finding lapack when looking at the extension modules, so I explicitly set `PATH=/bin:/usr/bin:/usr/lib/lapack` when looking at the extension modules and running the tests.
It's still failing, but it's starting to run the tests first, so I think it's just regular failure now (as opposed to CI failures not reproducible locally).  Unfortunately, the failure is a new segfault I need to track down.

Does anyone want me to rebase and squash some of the comments together or make the commit messages numpy-style (add TST: at start)?  I've recently had a couple rounds of "try to do something, then account for differences between local and CI systems"

**Comment by charris** [comment]
> Does anyone want me to rebase and squash some of the comments [commits?]

That would be good and you might as well rewrite the commit message(s) while doing so. The tests are passing now, do you think this is about ready?

**Comment by charris** [comment]
@mattip @seberg Ready for review.

File: .github/workflows/cygwin.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+          do
+              echo ${name}
+              python3.7 -c "import "$(echo ${name} | sed -E -e "s/\/+(home|usr).*?site-packages\/+//g" -e "s/\//./g" -e "s/.cpython-3.m?-x86(_64)?-cygwin.dll//g")
+          done
+          '
+          dash -c "tr -d '\r' <list_dlls_dos.sh >list_dlls_unix.sh"
+          bash -c "PATH=/bin:/usr/bin:/usr/lib/lapack bash list_dlls_unix.sh"
+      - name: Run NumPy test suite
+        run: |
+          dash -c "mkdir empty"
```
Comment (charris) [inline .github/workflows/cygwin.yml:66]: Is `dash` correct?

File: .github/workflows/cygwin.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+          dll_list=$(for name in $(python3.7 -m pip show -f numpy | grep -F .dll); do echo ${site_packages}/${name}; done)
+          ls -l ${dll_list}
+          file ${dll_list}
+          ldd ${dll_list}
+          cygcheck ${dll_list}
+          cd dist/
+          for name in ${dll_list};
+          do
+              echo ${name}
+              python3.7 -c "import "$(echo ${name} | sed -E -e "s/\/+(home|usr).*?site-packages\/+//g" -e "s/\//./g" -e "s/.cpython-3.m?-x86(_64)?-cygwin.dll//g")
```
Comment (charris) [inline .github/workflows/cygwin.yml:59]: Is it possible to break this line (and similar) with `\`?

File: .github/workflows/cygwin.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      - uses: actions/checkout@v2
+        with:
+          submodules: recursive
+          fetch-depth: 0
+      - name: Install Cygwin
+        uses: egor-tensin/setup-cygwin@v3
+        with:
+          platform: x64
+          install-dir: 'C:\tools\cygwin'
+          packages: python37-devel python37-pytest python37-hypothesis python37-cython python37-pip python37-wheel python37-setuptools liblapack-devel libopenblas gcc-fortran git
```
Comment (charris) [inline .github/workflows/cygwin.yml:22]: It should be possible to break this line with `packages: >`. See https://stackoverflow.com/questions/3790454/how-do-i-break-a-string-in-yaml-over-multiple-lines.

File: .github/workflows/cygwin.yml (hunk 1 above)
Comment (DWesl) [reply .github/workflows/cygwin.yml:66]: [Dash](http://gondor.apana.org.au/~herbert/dash/) is an alternate implementation of the Bourne shell.  It is faster than bash on Cygwin for many applications.  The website says it tries to be small and fast, and being small will help make `fork()` faster on Cygwin.  

I can change over to bash if you prefer, or let PowerShell execute that line since there's no files being written in that directory.

File: .github/workflows/cygwin.yml (hunk 3 above)
Comment (DWesl) [reply .github/workflows/cygwin.yml]: Done

File: .github/workflows/cygwin.yml (hunk 1 above)
Comment (charris) [reply .github/workflows/cygwin.yml:75]: No problem, I just thought it might be a typo. I've used dash during failed linux boots,

Comment (DWesl) [reply .github/workflows/cygwin.yml:75]: I added `dash` to the requirements to make sure it's available here.

File: .github/workflows/cygwin.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          cd dist/
+          for name in ${dll_list};
+          do
+              echo ${name}
+              python3.7 -c "import "$(echo ${name} | \
+                  sed -E \
+                  -e "s/\/+(home|usr).*?site-packages\/+//g" -e "s/\//./g" \
+                  -e "s/.cpython-3.m?-x86(_64)?-cygwin.dll//g")
+          done
+          '
```
Comment (charris) [inline .github/workflows/cygwin.yml:70]: Stray?

File: .github/workflows/cygwin.yml (hunk 2 above)
Comment (DWesl) [reply .github/workflows/cygwin.yml]: Some places, yes, other places, not easily.

I could try using `runtests.py` again; that should work now and would be shorter than the current `np.test()` line

File: .github/workflows/cygwin.yml (hunk 4 above)
Comment (charris) [reply .github/workflows/cygwin.yml:89]: I was wondering about `'`.

Comment (charris) [reply .github/workflows/cygwin.yml:89]: But now I see it starts way up above :)

Comment (DWesl) [reply .github/workflows/cygwin.yml:89]: Yes, that's the end of the extension-module-checking script.  Should I split that into a separate file somewhere rather than writing it out each time?

Comment (charris) [reply .github/workflows/cygwin.yml:89]: I think the first thing is to get the test passing again, we can worry about further cleanups after that. I'm beginning to regret the original suggestion :) But it does look _much_ nicer now.

**Comment by DWesl** [comment]
This is now failing within the actual NumPy test suite, somewhere inside `numpy/core/tests/test_simd.py`.  

If I use `runtests.py` rather than `np.test()` to launch the tests, I can split the tests in that file from the rest of the test suite.  The majority of the test suite has no segfaults and can run quickly using the default runner, while the simd tests do have segfaults and need to be run with pytest-forked to provide details on which tests are failing, especially if the tests are parameterized.

Should I switch to using `runtests.py` again?  Should I split out the tests in `numpy.core.tests.test_simd` from the rest of the suite so we actually get an error count for most of the code base?

**Comment by charris** [comment]
I don't know what you should do. Were the actual tests running before? Runtests is probably the best interface for debugging, just use `-n` (test installed wheel) and `-vv` (more verbosity).

**Comment by charris** [comment]
Might be nice to know what compiler version Cygwin is using.

**Comment by DWesl** [comment]
The compiler should be GCC 10.2.0. 

 I forgot to change the main branch name from `master` to `main`, so there are a decent number of commits where the workflow wasn't run on the PR.  There are still the test failures that have been around for the last few NumPy versions my other PR is aimed at fixing, plus the segfaults in test_simd that I didn't see in older versions.

**Comment by charris** [comment]
The segfault is in ../../core/tests/test_simd.py::Test_SIMD_BOOL_256_FMA3__AVX2_b8::test_operators_logical. Maybe @seiko2plus can help.

**Comment by seiko2plus** [comment]
@charris, I will dig into it

**Comment by seiko2plus** [comment]
> ../../core/tests/test_simd.py::Test_SIMD_BOOL_256_FMA3__AVX2_b8::test_operators_logical

```Bash

│   0x44e2a2e29 <PySIMDVector_AsData+41>    callq  0x44e3a39e8 <PyObject_IsInstance>                                                                                                                             │
│   0x44e2a2e2e <PySIMDVector_AsData+46>    test   %eax,%eax                                                                                                                                                     │
│   0x44e2a2e30 <PySIMDVector_AsData+48>    je     0x44e2a2ea0 <PySIMDVector_AsData+160>                                                                                                                         │
│   0x44e2a2e32 <PySIMDVector_AsData+50>    mov    0x10(%rsi),%eax                                                                                                                                               │
│   0x44e2a2e35 <PySIMDVector_AsData+53>    cmp    %ebx,%eax                                                                                                                                                     │
│   0x44e2a2e37 <PySIMDVector_AsData+55>    jne    0x44e2a2e60 <PySIMDVector_AsData+96>                                                                                                                          │
│   0x44e2a2e39 <PySIMDVector_AsData+57>    vmovdqu 0x20(%rsi),%xmm1                                                                                                                                             │
│   0x44e2a2e3e <PySIMDVector_AsData+62>    vinserti128 $0x1,0x30(%rsi),%ymm1,%ymm0                                                                                                                              │
│  >0x44e2a2e45 <PySIMDVector_AsData+69>    vmovdqa %ymm0,(%r12)  
```
GCC tries to aligned load local variable of 256-bit vector from non-aligned 256-bit stack pointer.
The bug has been reported multiple times and only limited for GCC on windows, see:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=49001
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=40838
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54412

It seems this bug will never be fixed and the only workaround is to avoid dealing with GCC ABI when
it comes to data types that requires alignment larger than 128-bit.
In other words force inlining any function that accepts or returning data types of __m256/__m512, or just pass or returns any variables by constant reference.
Another aggressive solution would be force disable AVX/AVX512 on GCC/Windows.

**Comment by charris** [comment]
@seiko2plus Thanks for the update. Sounds like disabling that for gcc on windows would be the way to go. We do produce 32 bit windows wheels using mingw, but I suppose avx is not a problem on that platform.  @mattip Do you know if we actually use gcc to produce windows wheels? I'm pretty sure we use MSVC for 64 bits.

**Comment by seiko2plus** [comment]
@charris, Disabling AVX may lead to a significant loss of performance, I would prefer to work under this BUG instead. all we need is to avoid exporting any functions that accept or return AVX registers, SIMD interface already force inlining all intrinsics. #18795 should pass the SIMD test.

**Comment by charris** [comment]
@seiko2plus  Sounds good.

**Comment by charris** [comment]
Close/reopen

**Comment by charris** [comment]
Different segfault in `../../core/tests/test_multiarray.py::TestArgmax::test_all `

**Comment by charris** [comment]
Let's try again.

**Comment by DWesl** [comment]
```
Fatal Python error: Segmentation fault

Current thread 0x0000000800000010 (most recent call first):
  File "/usr/lib/python3.7/site-packages/numpy/core/_methods.py", line 40 in _amax
  File "/usr/lib/python3.7/site-packages/numpy/core/tests/test_multiarray.py", line 4265 in test_all
```

I don't know NumPy internals well enough to trace that back to its source unless the problem is just that this function needs  `CYG_FINLINE`:
https://github.com/numpy/numpy/blob/914407d51b878bf7bf34dbd8dd72cc2dbc428673/numpy/core/src/umath/loops.h.src#L215-L225

**Comment by DWesl** [comment]
For more fun, running the test suite for NumPy `main` completes with only the expected test failures (two new ones since 1.16 or so) and no segfaults, so this seems to be something peculiar to the runner.

**Comment by charris** [comment]
Yay, the segfaults are gone. Thanks Sayed. @DWesl Now to figure out why the tests are failing. I notice that a bunch of them are with complex256, not sure why. Does Cygwin actually have float128 on windows? You might try block listing  some Cygwin functions in `numpy/core/src/common/npy_config.h`.

**Comment by DWesl** [comment]
Cygwin has float128 and complex256, but I think the C math functions got pulled from somewhat buggy versions in Mingw-w64 a while back, and haven't been updated as the upstream versions got fixed unless someone points out that would be useful.  About half those failures are related to `hypotl(inf, nan) = cabsl(inf + nan*j) = nan`, in either order and with either sign for the infinity, and one of the others is `hypotl(LDBL_MAX, 0.0L) = inf`.

I summarized the failures a few months ago at the start of #18308: should I cherry-pick the relevant commits over to this PR? I think cfa9b4e380d6cd396455f63a57891bec8d4bfe48 does npy_config.h stuff, though I could go through and check which functions are still relevant.

**Comment by seiko2plus** [comment]
@charris,
> Different segfault in ../../core/tests/test_multiarray.py::TestArgmax::test_all

segfault caused due to the same GCC/WIN64 bug but this time with `AVX512`, here the error report from Intel SDE

```Bash
>>> a = np.random.normal(0, 1, (4, 5, 6, 7, 8))
>>> for i in range(a.ndim):
...             amax = a.max(i)

SDE ERROR:  TID: 0 executed instruction with an unaligned memory reference to address 0xffffb5d0 INSTR: 0x46e8d7706: IFORM: VMOVAPD_MEMf64_MASKmskw_ZMMf64_AVX512 :: vmovapd zmmword ptr [rax], zmm0
        IMAGE:    \\?\C:\cygwin64\lib\python3.7\site-packages\numpy\core\_multiarray_umath.cpython-37m-x86_64-cygwin.dll
        FUNCTION: .text
        FUNCTION ADDR: 0x46e8b8a60
```
@DWesl,
> (two new ones since 1.16 or so) and no segfaults, so this seems to be something peculiar to the runner.

That because segfaults will only appears on X86 CPUs with AVX512 support


@charris,
> Yay, the segfaults are gone. Thanks Sayed

My pleasure, the segfaults should gone but I'm not sure yet since the current CI/Worker doesn't support `AVX512`.

**Comment by charris** [comment]
close/reopen

**Comment by DWesl** [comment]
I told NumPy about the functions mentioned in the test failures, which got rid of some of the failures.  Blocking `hypotl` as well as `cabsl` might eliminate most of the `np.complex256` failures.  I'm not sure why blocking `casinh` and friends leaves the failures in `test_loss_of_precision`, especially since `casin` seems like it wouldn't fail this way.  

I'm not sure what direction this PR should go from here, since I don't understand why the remaining failures are still there.

**Comment by charris** [comment]
close/reopen

**Comment by DWesl** [comment]
So, fun thing happened when trying to add 
```C
#undef HAVE_CABSL
#undef HAVE_HYPOTL
```
to the current version: with the above code in `npy_config.h`, I get segfaults GDB says are in `PySIMDVector_AsData` called from `simd_arg_converter`.  Without them (that is, with the current version on GitHub), I get segfaults GDB says are in `simd_arg_to_obj` called from `simd.intrin_setall_u8`.  GDB says there's no symbol table available, so I'm not sure how accurate those are.  How would I go about debugging this further?

This machine has AVX2 but not AVX512, from what I can tell.

**Comment by charris** [comment]
That's interesting. The fallback cabsl calls hypotl, which falls back to double precision and calls sqrt, which is where I think SIMD creeps in. How does Cygwin treat long double? Windows defines it as double, I wonder if that is an option for Cygwin?

**Comment by DWesl** [comment]
I should probably mention: pytest says the crash is in `numpy.core.tests.test_simd.test_operators_logical`.  I just pushed the commit marking `cabsl` and `hypotl` for replacement, which should take the number of CI test failures down a bit.  If I run
```bash
PATH=/bin:/usr/bin:/usr/lib/lapack NPY_BLAS_ORDER=blas NPY_LAPACK_ORDER=lapack python runtests.py -- --ignore=numpy/core/tests/test_simd.py
```
the test suite completes without segfaults and only nine failures.  
- `cpowl` doesn't report overflow, even when `cpowl` and `powl` are marked for replacement.
- `cabsl` reports overflow in some cases it should complete (with `cabsl` and `hypotl` marked for replacement)
- `csqrt` and `csqrtf` don't do branch cuts properly when marked for replacement
- `casinhf`, `casinh`, and `casinhl` lose precision for small arguments
- `SVD` doesn't behave like it should
- CFFI can't find `numpy.random._generator.random_standard_normal_fill`

Cygwin uses the Newlib implementations for `long double` functions by default.  Newlib seems targeted at systems without much memory, so the `long double` implementations there tend to be recent and simple.  On some occasions, people have made the case that Cygwin should have better implementations, at which point the maintainers often grab the ones from the Mingw-w64 project.  

I have no idea how one would go about disabling long double for a NumPy build, but it would likely get the CI passing if you want to do that.

**Comment by DWesl** [comment]
It's down to four failures now:
- `numpy.core.tests.test_datetime.TestDateTime.test_datetime_as_string_timezone`: Expected `01:30-05`, Got `06:30+00`
- `numpy.core.tests.test_scalarmath.TestAbs.test_builtin_abs[complex256]`: `abs(HUGE+0j) == inf`, not HUGE
- `numpy.core.tests.test_scalarmath.TestAbs.test_numpy_abs[complex256]`: Encountered overflow in absolute
- `numpy.linalg.tests.test_linalg.TestCond.test_nan`: computing SVD of a matrix with a `nan` fails to converge

I don't know why these two `cabsl` tests still fail when all the other ones pass, and I don't know enough about the other two to know why those are failing.  Any ideas?

I've made sure pytz is installed on the runner.  Is NumPy expecting another timezone package to be available as well?

**Comment by DWesl** [comment]
Down to three errors now.  I can find the backup definition for `npy_cabsl` in terms of `npy_hypotl`: 
https://github.com/numpy/numpy/blob/0f7f313d847e8e36e9fb3c4e576619a3af3bfe56/numpy/core/src/npymath/npy_math_complex.c.src#L121-L127
but I can only find a definition for `npy_hypot`, not `npy_hypotl`:
https://github.com/numpy/numpy/blob/0f7f313d847e8e36e9fb3c4e576619a3af3bfe56/numpy/core/src/npymath/npy_math_internal.h.src#L191-L219
Where is `npy_longdouble npy_hypotl(npy_longdouble, npy_longdouble)` defined if `HAVE_HYPOTL` is not `#define`d?

**Comment by DWesl** [comment]
Found the `npy_hypotl` implementation:
https://github.com/numpy/numpy/blob/bb745921ac6c27d6caf9b76663668e5b9f28abac/numpy/core/src/npymath/npy_math_internal.h.src#L400-L413
It defaults to `npy_hypot`.  Since the tests for `numpy.abs(numpy.clongdouble)` require long double precision and `npy_cabsl` falls back to `npy_hypotl` if the former is not defined, I took the liberty of changing `npy_math_internal.h.src` to repeat the definition of `npy_hypot` with the new types instead:
https://github.com/numpy/numpy/blob/2328404ca5e114e3881184a1e5eee6fba2f739b4/numpy/core/src/npymath/npy_math_internal.h.src#L191-L226
For some reason, this only fixes the `np.abs(np.clongdouble)` tests, not the `abs(np.clongdouble)` ones, at least locally.

**Comment by DWesl** [comment]
Should I leave the redefinition of `npy_hypot[fl]` in this PR, or should I move it to a different PR?

I should perhaps mention the two failing checks appear to be the two CircleCI builds that have been failing intermittently over the course of this PR.

**Comment by DWesl** [comment]
I went back to the old definition of `npy_hypotl` in terms of `npy_hypot` and marked the `npy_cabsl` tests xfail, since those expect `long double` precision rather than merely double.

**Comment by DWesl** [comment]
Is there a preference on whether I change the `cabsl` test to match the current fallback implementation (`npy_hypotl`->`npy_hypot`) or change the `hypotl` fallback implementation to pass the current `cabsl` test?  I currently have the test matching the implementation, but I can go back to having the implementation match the test.  

The CI failures seem unrelated.

**Comment by charris** [comment]
The circleci failure should be fixed by a rebase. You might want to merge some of the commits first, though, otherwise you could end up fixing any merge failures in all 38 commits :) Let me try a close/reopen first.

**Comment by DWesl** [comment]
The CircleCI failures are still there, so here's one rebase.  I merged many of the commits where I was trying to figure out YAML syntax and how that got passed to PowerShell, but only ran into merge conflicts with other people's code in the last two commits.

Did the "SIMD: Force inlining all functions that accept AVX registers" commit get rolled back?  I think I grabbed that from `main` so the tests would stop segfaulting.

**Comment by charris** [comment]
> Should I try reverting that commit and seeing if the tests still pass without segfaults?

You could try. I don't see an indication that it was reverted, I do know that we have been inlining more SIMD functions because some compilers fail to do so, leading to problems.  You could just check out the relevant files from main for such a check,
```
git checkout main <filename>
``` 
which will put them in the index for a commit (IIRC) and also overwrite any local changes when merged. If fixes are needed after that they should probably go into a separate PR.

**Comment by DWesl** [comment]
> > Should I try reverting that commit and seeing if the tests still pass without segfaults?
> 
> You could try. I don't see an indication that it was reverted, I do know that we have been inlining more SIMD functions because some compilers fail to do so, leading to problems. You could just check out the relevant files from main for such a check,
> 
> ```
> git checkout main <filename>
> ```
> 
> which will put them in the index for a commit (IIRC) and also overwrite any local changes when merged. If fixes are needed after that they should probably go into a separate PR.

I did that, and the tests that had been segfaulting still pass, so I pushed it up for CI.

**Comment by DWesl** [comment]
That seems to have removed too many function inline annotations.  The Cygwin test reports a segfault while executing

> File "/usr/lib/python3.7/site-packages/numpy/core/tests/test_api.py", line 20 in test_array_array
> > ```python
> >     ones11 = np.ones((1, 1), np.float64)
> >     assert_equal(np.array(ones11, dtype=np.float64), ones11)
> > ```

I suspect one of the annotations was for an SIMD equality comparison loop.

**Comment by charris** [comment]
The next step would be to make a PR adding the annotations that you think are needed, then we can ask @seiko2plus for a review.

**Comment by charris** [comment]
It could be that the relevant changes were lost unintentionally.

**Comment by DWesl** [comment]
My suspicion is they were lost in merge conflict resolution.  I can cherry-pick the original commit again as a new PR.

**Comment by DWesl** [comment]
I added that commit in a PR as #19529.

After digging way back through the pre-rebase commits, I found the original as 05598082761d172f9a5c37298f57cb68ece73386, which seems to have been originally authored on this branch.  There was a different commit that I stole from `main`, which git did indeed drop from the rebase.

**Comment by DWesl** [comment]
Rebasing to include that commit again seems to have gotten the tests working again.  Anything else I need to do?

One of the changes since the original version of this PR was to add `fmodl` to the list of functions that default to `double` precision if `long double` is not available, so I'm fairly confident leaving `hypotl` in that list and changing the test to match was the right decision.

File: numpy/core/tests/test_scalarmath.py [hunk 5]
```diff
@@ -678,11 +678,35 @@ def _test_abs_func(self, absfunc, test_dtype):
...
+                dtype,
+                marks=pytest.mark.xfail(
+                    sys.platform == "cygwin" and dtype == np.clongdouble,
+                    reason="npy_cabsl calls npy_hypotl, which is npy_hypot",
+                ),
+            )
+            for dtype in floating_types + complex_floating_types
+        ],
+    )
     def test_builtin_abs(self, dtype):
```
Comment (charris) [inline numpy/core/tests/test_scalarmath.py:694]: A cleaner way to do this is to raise skip inside the function. Something like
```
def test_builtin_abs(self, dtype):
    if dtype == np.clongdouble and sys.platform == "cygwin" :
        pytest.skip("absl is computed in double precision on cygwin")
    ...
```
You could also use `pytest.xfail` instead of `pytest.skip`.

**Comment by charris** [comment]
Could move some of the github workflow into a `tools/cygwin_test.sh`, see `.github/actions/action.yml` for an example.  But that can wait.

File: numpy/core/tests/test_scalarmath.py (hunk 5 above)
Comment (DWesl) [reply numpy/core/tests/test_scalarmath.py:682]: Done, here and for the next test.

**Comment by charris** [comment]
In it goes, thanks @DWesl .
